### PR TITLE
Add torch.Tensor.as_subclass method.

### PR DESCRIPTION
**Pull request by hameerabbasi**
This is according to pytorch/rfcs#3.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
The CI failures seem to be due to network disruption.

**Review by rgommers — CHANGES_REQUESTED** [review]
First thoughts:

1. This needs a good set of tests and docs
2. Does it make sense to merge this without the other changes in the RFC? If not, using `ghstack` to make the related PRs seems like the way to go.

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -539,7 +539,7 @@ static PyMappingMethods THPVariable_as_mapping = {
 };
 
 static PyMethodDef extra_methods[] = {
-  {"_make_subclass", (PyCFunction)(void(*)(void))THPVariable_make_subclass, METH_STATIC | METH_VARARGS | METH_KEYWORDS, nullptr},
+  {"make_subclass", (PyCFunction)THPVariable_make_subclass, METH_STATIC | METH_VARARGS | METH_KEYWORDS, nullptr},
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp]: I would be in favor of keeping both here. I am pretty sure some people use this outside our codebase.

Comment (hameerabbasi) [reply torch/csrc/autograd/python_variable.cpp]: Done.

**Comment by rgommers** [comment]
@hameerabbasi I just remembered that in https://github.com/pytorch/pytorch/issues/28361 the agreed upon name was `as_subclass`, and there's some fastai code already that assumed that. Not that that is then set in stone, however `as_subclass` is probably a clearer name than `make_subclass`.

**Comment by hameerabbasi** [comment]
> and there's some fastai code already that assumed that

https://github.com/fastai/fastai2/search?q=as_subclass&unscoped_q=as_subclass

True. The RFC and this PR will both need updating.

File: test/test_torch.py [hunk 2]
```diff
@@ -1208,6 +1208,26 @@ def test_dtype_out_match(self):
         d = torch.autograd.Variable(torch.DoubleTensor(2, 3))
         self.assertRaises(RuntimeError, lambda: torch.zeros((2, 3), out=d, dtype=torch.float32))
 
+    def test_make_subclass(self):
+        class SubTensor(torch.Tensor):
+            pass
```
Comment (ezyang) [inline test/test_torch.py]: I think it's important to also test what happens to member variables on the class, as this is a very important part of the input contract.

File: test/test_torch.py [hunk 3]
```diff
@@ -1208,6 +1208,26 @@ def test_dtype_out_match(self):
...
+
+        t0 = torch.Tensor(0)
+        t1 = torch.Tensor([1, 2])
+        t2 = torch.Tensor([[3, 4], [5, 6]])
```
Comment (ezyang) [inline test/test_torch.py]: nit: for tests, try to use the new style constructor `torch.tensor` rather than the legacy constructor

File: torch/csrc/autograd/python_variable.cpp [hunk 4]
```diff
@@ -144,6 +144,32 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
 }
 
 // Instantiates a subclass of torch.Tensor. Used by nn.Parameter()
+static PyObject* THPVariable_as_subclass(THPVariable* self, PyObject* args, PyObject* kwargs) {
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp]: Uhh, is this a straight copy paste of `make_subclass`? Can we avoid doing that?

Comment (hameerabbasi) [reply torch/csrc/autograd/python_variable.cpp]: It isn't, it's missing the detach and the `requires_grad` calls, and the signature is different. That's for a reason: We want this to be a straight view, not a detached-and-re-attached view as `Parameter` requires.

Comment (ezyang) [reply torch/csrc/autograd/python_variable.cpp]: So, you don't want the comment above right?

```
// Instantiates a subclass of torch.Tensor. Used by nn.Parameter()
```

File: torch/csrc/autograd/python_variable.cpp [hunk 5]
```diff
@@ -144,6 +144,32 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
...
+    "as_subclass(PyObject* cls)",
+  });
+  ParsedArgs<2> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  PyObject* cls = r.pyobject(0);
+  if (!PyType_Check(cls)) {
+    throw TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
+  }
+  auto data = as_variable_ref(self->cdata);
+  // We set `data`'s `allow_tensor_metadata_change` to true here, because we want to
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp]: This is a completely new function right? Why is there this comment about backwards compatibility?

Comment (hameerabbasi) [reply torch/csrc/autograd/python_variable.cpp]: Because it was a copypasta before https://github.com/pytorch/pytorch/pull/34369#issuecomment-599635556.

File: torch/csrc/autograd/python_variable.cpp [hunk 6]
```diff
@@ -144,6 +144,32 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
...
+  // allow the following use case for backward compatibility:
+  //
+  // ```python
+  // rnn = torch.nn.RNN(100, 100, 2)
+  // # The following calls `torch._cudnn_rnn_flatten_weight(rnn._flat_weights, ...)`,
+  // # which changes storage of `rnn`'s weights in-place
+  // rnn.flatten_parameters()
+  // ```
+  data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
+  return THPVariable_NewWithVar((PyTypeObject*)cls, std::move(data));
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp]: I think your modifications here make this unsound. `data` corresponds directly to the passed in Python object (not a new alias of it, as was the case in `THPVariable_make_subclass`, and then you call `THPVariable_NewWithVar` which has this invariant:

```
// Creates a new Python object for a Variable. The Variable must not already
// have a PyObject* associated with it.
static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
```

This invariant is violated: you already have a PyObject, the one the variable came in with.

File: test/test_torch.py [hunk 7]
```diff
@@ -1208,6 +1208,30 @@ def test_dtype_out_match(self):
         d = torch.autograd.Variable(torch.DoubleTensor(2, 3))
         self.assertRaises(RuntimeError, lambda: torch.zeros((2, 3), out=d, dtype=torch.float32))
 
+    def test_make_subclass(self):
```
Comment (ezyang) [inline test/test_torch.py]: nit: `test_as_subclass`

File: torch/csrc/autograd/python_variable.cpp (hunk 6 above)
Comment (hameerabbasi) [reply torch/csrc/autograd/python_variable.cpp]: Would there be a way to make an alias that wouldn't violate this invariant and keep its position in the computation graph?

Comment (hameerabbasi) [reply torch/csrc/autograd/python_variable.cpp]: Had a quick chat with @ezyang: We agreed that `.alias()` was the way to go, it'll keep the new object attached to autograd as an alias of the original, so autograd should continue to function.

File: torch/_tensor_docs.py [hunk 8]
```diff
@@ -3576,3 +3576,12 @@ def callable(a, b) -> number
 If ``n`` is the number of dimensions in ``x``,
 ``x.T`` is equivalent to ``x.permute(n-1, n-2, ..., 0)``.
 """)
+
+add_docstr_all('as_subclass',
+               r"""
+as_subclass(cls) -> Tensor
+
+Makes a ``cls`` instance with the same data pointer as ``self``. Changes
+in the output mirror changes in ``data``. ``cls`` must be a subclass of
```
Comment (albanD) [inline torch/_tensor_docs.py]: Not sure what do you mean by `data` here? Does it reference the `.data` field of Tensors?
Isn't this closer to the `.detach()` semantic?

Comment (hameerabbasi) [reply torch/_tensor_docs.py]: Should have been `self`. It stays connected to the autograd graph.

Comment (hameerabbasi) [reply torch/_tensor_docs.py]: Please read the newer version to see if it's clearer.

File: test/test_torch.py [hunk 9]
```diff
@@ -1208,6 +1208,30 @@ def test_dtype_out_match(self):
...
+        self.assertTrue(type(s2) is SubTensor)
+
+        self.assertEqual(t0, s0)
+        self.assertEqual(t1, s1)
+        self.assertEqual(t2, s2)
+
+        self.assertTrue(s0.member_var is SubTensor.member_var)
+        self.assertTrue(s1.member_var is SubTensor.member_var)
+        self.assertTrue(s2.member_var is SubTensor.member_var)
+
```
Comment (ezyang) [inline test/test_torch.py]: You're maintaining the autograd graph, so I'd expect a test showing that autograd still works through subclassed tensors.

File: torch/csrc/autograd/python_variable.cpp [hunk 10]
```diff
@@ -143,7 +143,24 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
...
+    "as_subclass(PyObject* cls)",
+  });
+  ParsedArgs<2> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  PyObject* cls = r.pyobject(0);
+  if (!PyType_Check(cls)) {
+    throw TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
+  }
+  auto data = as_variable_ref(self->cdata.alias());
+  data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp]: I don't think you need this line anymore, `alias()` should allow metadata change automatically. You can test for this by running `resize_` on the result.

File: torch/csrc/autograd/python_variable.cpp [hunk 11]
```diff
@@ -143,7 +143,24 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
...
+  static PythonArgParser parser({
+    "as_subclass(PyObject* cls)",
+  });
+  ParsedArgs<2> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  PyObject* cls = r.pyobject(0);
+  if (!PyType_Check(cls)) {
+    throw TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
+  }
+  auto data = as_variable_ref(self->cdata.alias());
```
Comment (ezyang) [inline torch/csrc/autograd/python_variable.cpp]: btw, `as_variable_ref` doesn't actually do anything anymore, so you don't need it :)

File: test/test_torch.py (hunk 9 above)
Comment (ezyang) [reply test/test_torch.py]: More generally, I'd spend a little more time thinking about what kinds of properties you want to test.

File: torch/_tensor_docs.py (hunk 8 above)
Comment (albanD) [reply torch/_tensor_docs.py]: Looks good thanks !

File: test/test_torch.py [hunk 12]
```diff
@@ -1209,6 +1209,51 @@ def test_dtype_out_match(self):
...
+        # Test that autograd is propagated.
+        t = torch.tensor(5, dtype=torch.float32, requires_grad=True)
+        t_ = torch.exp(t)
+
+        s_ = t_.as_subclass(SubTensor)
+        t_.backward()
+
+        self.assertEqual(t_.grad, s_.grad)
```
Comment (ezyang) [inline test/test_torch.py]: This is actually really surprising! Does this actually pass?!

Comment (ezyang) [reply test/test_torch.py]: Are you sure these aren't just None in both cases...

**Comment by ezyang** [comment]
I don't understand the changes you made to the autograd test. Please explain in more words, or phone me on VC when you have a chance.

**Review by ezyang — APPROVED** [review]

**Comment by hameerabbasi** [comment]
Is there anything more to be done here?

**Comment by ezyang** [comment]
no, it's my bad. Landing.
